Begin work on Ruby Guide with Rubocop template

monfresh opened the pull request:
**Why**: To get a new Ruby section started and a discussion on our
recommended Rubocop configuration.

monfresh commented:
cc @jessieay @pkarman @harrisj @afeld @dogweather

jessieay commented on README.md:
```diff
@@ -7,6 +7,7 @@ A set of guidelines and best practices for an awesome engineering team
 * [Project Setup](/project_setup)
 * [Example Workflows](/example_workflows)
 * [Architecture Reviews](/architecture_reviews)
+* [Ruby Guide](/ruby_guide)
```
what about just calling the path `ruby`

monfresh replied:
Sounds good to me. I'll update it.

harrisj commented:
It begins! Awesome... I wonder if we can look at the rubocop configs for various projects to see the differences from the stock rubocop.yml we generally agree about and those we will fight about. I would love to have a central consistent style guide though

harrisj commented:
Where is this rubocop.yml from?

jessieay commented:
This looks good to me! Also want to add something on setting up code climate for rubocop and other linters since I found the process to be a bit confusing (added to Dolores today)

monfresh commented:
This Rubocop config is from the Identity project (and save-ferris before that). It's basically the one I use and have suggested to my teammates. The only objection that was ever raised was the line length, which was originally at 80 and we bumped it up to 100.

jessieay commented:
very excited about this! C2 and Micropurchase have rubocop files that are pretty different from this so comparing will be helpful

monfresh commented:
@jessieay Yes, I will add a Code Climate config as well and instructions. My goal is to have all of this automated. You press a button and you get a shiny new Rails app with all the recommended gems, static analysis, and Travis configured.

sbc100 commented on ruby_guide/.rubocop.yml:
```diff
@@ -0,0 +1,101 @@
+  - spec/**/*
+
+Metrics/LineLength:
+  Description: Limit lines to 80 characters.
```
This is inconsistent with the line below..

monfresh replied:
The description is from the Rubocop default config. I think it's helpful to know what the default setting is.

sbc100 commented on ruby_guide/.rubocop.yml:
```diff
@@ -0,0 +1,101 @@
+  - conditionals
+
+Style/Documentation:
+  Description: 'Document classes and non-namespace modules.'
```
This is a getting a bit meta, but is there a reason this is the only description quotes?

monfresh replied:
I copied and pasted from Rubocop: https://github.com/bbatsov/rubocop/blob/master/config/enabled.yml#L179-L184

dogweather commented:
Sweet. I like the `yml` file you added, because it enforces most everything, especially ruby idioms.

Does it need an explicit 

```yml
Rails:
  Enabled: true
```

?

jessieay commented on ruby_guide/.rubocop.yml:
```diff
@@ -0,0 +1,101 @@
+
+Metrics/MethodLength:
+  Description: Avoid methods longer than 10 lines of code.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#short-methods
```
I do love having a link for each rule! sometimes very hard to tell what something means without a longer description

harrisj commented on ruby_guide/.rubocop.yml:
```diff
@@ -0,0 +1,101 @@
+    - '**/Rakefile'
+  Exclude:
+    - 'bin/**/*'
+    - 'db/schema.rb'
```
Should we move the `spec/**/*` exclusion up here?

harrisj replied:
Ah never mind, it would be useful in some specs

harrisj commented:
So, here is a question I was wondering about: I know we technically only have to define Rubocop rules we might change from the default config, but do you think we should copy over the entire default config and make our changes to that so it's clear what all our rules are? If we did do that though, we'd have to be able to track with Rubocop changes/adds new defaults too.

sbc100 commented:
I would suggest that we only define how we differ from upstream style, and link to upstream for the defaults.   Ideally each is difference from upstream norms would have a good justification too (in an ideal world we would not differ and simply point to upstream for all style issues).

For example, chromium defines its style only in how it differs from Google's style guide: https://www.chromium.org/developers/coding-style

monfresh commented:
I agree that we should only keep the cops that differ from the defaults. I will add a link to the default config.

harrisj commented:
One thing I've also learned while looking our project's rubocop configs is that rubocop supports a special `inherit_from` directive if you ever need to inherit/merge from other files. You can [see it in action in rubocop's config](https://github.com/bbatsov/rubocop/blob/master/config/default.yml)

harrisj commented:
So, I wrote a program to pull in all of our rubocop configs and compare what we are overriding in various projects. The [results can be viewed by 18F staff here](https://docs.google.com/a/gsa.gov/spreadsheets/d/10-6m-9OMDixfHjsgyEYwmqzFQprmoMG9ySlp5S1j42g/edit?usp=sharing). It seems like the biggest variance are the `Metrics/` styles. We could potentially define general rules for other Style things and let projects override the Metrics with their own settings if we don't want to control that.

monfresh commented:
Thanks for those results! One thing I noticed is that some projects disable a cop entirely instead of choosing one style over another. If a cop provides 2 choices, the cop should stay enabled and one choice should be made. This ensures consistency. For example, the `Style/AlignParameters` cop has been disabled in many projects. That means that sometimes the alignment could be with the first parameter, and other times with indentation. That is wrong IMO. You should pick one and stick to it.

monfresh commented:
By the way, I updated this PR with the feedback so far. I changed the endpoint to `ruby` (used to be `ruby_guide`), and added links to the default, enabled, and disabled official Rubocop configs.

jessieay commented:
This looks good to me! Meta question (probably should be written down in a separate PR): under what conditions should we merge PRs to guides? seems like bar should be higher than with a regular code PR since these are related to guidelines for the whole team to follow/consider...

cmc333333 commented:
Agreed that we want a semi high bar. I think in this case we can hold a quick up/down vote in #ruby (referencing it in #dev) once this PR's ready.

monfresh commented:
For those contested cops (from Jacob's spreadsheet) that provide different options, I will add them to this proposed config so that people who choose to copy it over to their projects can more easily change the option. After that, I think this is ready for a vote.

monfresh commented:
I just pushed an updated config that includes the Style cops most projects have disabled. I purposefully removed the `Enabled` key as a way of indicating that this cop should always be enabled, but that you can choose your preferred style. I also added this paragraph to the README:
```
Whenever a `Style` Rubocop setting provides multiple options, at least one
option must be chosen. A cop that supports different styles must never be
disabled outright. The point is to pick one style and use it consistently.
```

dogweather commented on ruby/README.md:
```diff
@@ -0,0 +1,24 @@
+
+Whenever a `Style` Rubocop setting provides multiple options, at least one
+option must be chosen. A cop that supports different styles must never be
+disabled outright. The point is to pick one style and use it consistently.
```
👍

harrisj commented:
This is cool by me. My only question is given the variance for the various `Metrics/` guides, do you think those should be specified in the default at all

harrisj commented:
Oops, I hit the wrong button. Sorry!

harrisj commented on ruby/.rubocop.yml:
```diff
@@ -0,0 +1,212 @@
+Style/SignalException:
+  Description: 'Checks for proper usage of fail and raise.'
+  StyleGuide: 'https://github.com/bbatsov/ruby-style-guide#prefer-raise-over-fail'
+  EnforcedStyle: only_raise
```
So we want to avoid the `fail` keyword entirely then?

monfresh replied:
I don't have strong feelings either. I've just followed the Guide for this. They seem to have reversed their rule. Before, they preferred `fail` for exceptions, now they recommend `raise`. The `semantic` option is not documented, so I'm not sure when it would be appropriate to use `fail`.

harrisj replied:
This is excellent and I'm very excited for this PR! Thank you again for
making this happen!

On Thu, Jun 30, 2016 at 10:56 AM, Moncef Belyamani <notifications@github.com
> wrote:

> In ruby/.rubocop.yml
> <https://github.com/18F/development-guide/pull/22#discussion_r69147951>:
>
> > +  # But it can be overridden by setting this parameter
> > +  IndentationWidth: ~
> > +
> > +Style/MultilineOperationIndentation:
> > +  EnforcedStyle: aligned
> > +  SupportedStyles:
> > +    - aligned
> > +    - indented
> > +  # By default, the indentation width from Style/IndentationWidth is used
> > +  # But it can be overridden by setting this parameter
> > +  IndentationWidth: ~
> > +
> > +Style/SignalException:
> > +  Description: 'Checks for proper usage of fail and raise.'
> > +  StyleGuide: 'https://github.com/bbatsov/ruby-style-guide#prefer-raise-over-fail'
> > +  EnforcedStyle: only_raise
>
> I don't have strong feelings either. I've just followed the Guide for
> this. They seem to have reversed their rule. Before, they preferred fail
> for exceptions, now they recommend raise. The semantic option is not
> documented, so I'm not sure when it would be appropriate to use fail.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/18F/development-guide/pull/22/files/b605d3b79f8e02563f915cf1ef7248090de12fe0#r69147951>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAH_Jr9iiK3457_nNFWBCnFmLRvTUeXks5qQ9kYgaJpZM4I-E10>
> .
>

harrisj commented on ruby/.rubocop.yml:
```diff
@@ -0,0 +1,212 @@
+Style/StringLiterals:
+  Description: Checks if uses of quotes match the configured preference.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#consistent-string-literals
+  EnforcedStyle: single_quotes
```
This is the rule that it should use `'abc'` not `"abc"` if there is no need for interpolation or special chars, right? How do you all feel about this one?

monfresh replied:
Yes, this makes it easy to distinguish between normal strings and interpolated strings.

afeld replied:
I have a somewhat unusual style for strings where I use single quotes for key-like things and double quotes for sentence-like things (which is where you would more likely have things like interpolation and special characters anyway)...therefore I usually disable it.

monfresh commented:
I don't see much variance for the `Metrics` cops. The only one that is heavily modified across projects is the `LineLength`, and I think 100 is a generous default setting.
The other one that a few projects have modified is the `MethodLength`. I think 10 lines is plenty for a method. We haven't had an issues sticking with that on the projects I've worked on.

jessieay commented on ruby/.rubocop.yml:
```diff
@@ -0,0 +1,212 @@
+
+# Checks the indentation of the first element in an array literal.
+Style/IndentArray:
+  # The value `special_inside_parentheses` means that array literals with
```
I really like the ` StyleGuide: 'https://github.com/bbatsov/ruby-style-guide#prefer-raise-over-fail'` type links here -- possible to get those for all rules rather than comments?

monfresh replied:
I copied and pasted these from the Rubocop repo. If a cop doesn't have a description, it's because one wasn't available in the Rubocop repo. I'm not sure if all cops have a matching entry in the Ruby Style Guide.

pkarman commented on ruby/.rubocop.yml:
```diff
@@ -0,0 +1,212 @@
+  - spec/**/*
+
+Metrics/LineLength:
+  Description: Limit lines to 80 characters.
```
should this match the `max` 100?

monfresh replied:
Not sure. I think it's helpful to know what the default Rubocop setting is. Perhaps the description can be reworded?

afeld replied:
I'd just leave out the descriptions, especially where the cop names are fairly self-explanatory.

pkarman commented:
`MethodLength` can be tricky when defining lists of things, since it is often most legible to put each item on a separate line. E.g.

```ruby

def list_of_allowed_params
  params.require(
    :foo,
    :bar,
    :baz,
    :color,
    :size,
    :shape,
    :condition,
    :mickey,
    :minnie,
    :donald
  )
end
```

As long as we are ok disabling the rules in specific places, I'm fine with the default of 10.

monfresh commented:
@pkarman Yes, those should be the exceptions, not the norm. Disabling a certain cop when there is no better option is perfectly fine.

cmc333333 commented:
Seems like a consensus to me
<img width="938" alt="screen shot 2016-06-30 at 5 28 19 pm" src="https://cloud.githubusercontent.com/assets/326918/16504865/1562bf82-3ee8-11e6-8363-20bc709db2c7.png">

afeld commented on ruby/.rubocop.yml:
```diff
@@ -0,0 +1,212 @@
+  Description: Limit lines to 80 characters.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#80-character-limits
+  Enabled: true
+  Max: 100
```
I feel pretty strongly that enforcing line length is a relic of a bygone era, and is more of a nuisance than a help these days, especially for sentence-like things (in strings, comments, and documentation). Not a blocker, but just letting you know that I plan to (propose) disabling it in my projects.

sbc100 replied:
I feel strongly the line line is important both is code and comments... we can bike shed for days on this subject :)

afeld commented on ruby/.rubocop.yml:
```diff
@@ -0,0 +1,212 @@
+# nil - warn on else with nil in it
+# both - warn on empty else and else with nil in it
+Style/EmptyElse:
+  EnforcedStyle: both
```
I'd advocate for `empty`, as I prefer being explicit about return values.

monfresh replied:
Can you give me an example where you had to specify `nil` in an `else` statement, and how it made the code more understandable?

afeld commented on ruby/.rubocop.yml:
```diff
@@ -0,0 +1,212 @@
+Style/IfUnlessModifier:
+  Description: Favor modifier if/unless usage when you have a single-line body.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#if-as-a-modifier
+  Enabled: true
```
The downside with enabling this is that many code coverage tools (e.g. `simplecov`) work by line, so they won't (or at least _didn't_) pick up whether the body(?) of a line with a modifier was hit or not.

monfresh replied:
Code Climate doesn't seem to have a problem counting modifier lines as covered. I'd have to test to see if when a modifier line is not covered, does it correctly flag it as not covered.

afeld commented:
I hope y'all appreciate how much I bit my tongue and minimized bikeshedding on this 😉

monfresh commented:
Should we vote on each cop in this PR?

cmc333333 commented:
Given the votes, I would strongly recommend merging as is and bring up individual points of contention in separate PRs.

monfresh commented:
Sounds good. Who wants to do the honors?

jessieay commented:
👏 👏 👏  thanks for submitting the PR, @monfresh ! and to everyone else who participated in this conversation. :)
